vexination/walker/src/server.rs

```diff
@@ -40,87 +49,138 @@ pub async fn run(
-                            discovered: DiscoveredAdvisory { url, .. },
-                            ..
-                        },
-                }) => {
-                    let name = url.path_segments().and_then(|s| s.last()).unwrap_or_else(|| url.path());
-                    match serde_json::from_slice::<csaf::Csaf>(&data) {
-                        Ok(doc) => match client
+    let total = Arc::new(AtomicUsize::default());
+    let duplicates: Arc<std::sync::Mutex<Duplicates>> = Default::default();
+    let errors: Arc<std::sync::Mutex<BTreeMap<DocumentKey, String>>> = Default::default();
+    let warnings: Arc<std::sync::Mutex<BTreeMap<DocumentKey, Vec<CheckError>>>> = Default::default();
+
+    {
+        let client = Arc::new(client.build()?);
+        let total = total.clone();
+        let duplicates = duplicates.clone();
+        let errors = errors.clone();
+        let warnings = warnings.clone();
+
+        let visitor = move |advisory: Result<
+            VerifiedAdvisory<ValidatedAdvisory, &'static str>,
+            VerificationError<ValidationError, ValidatedAdvisory>,
+        >| {
+            (*total).fetch_add(1, Ordering::Release);
+
+            let errors = errors.clone();
+            let warnings = warnings.clone();
+            let client = client.clone();
+            let sink = sink.clone();
+            let provider = provider.clone();
+            async move {
+                let adv = match advisory {
+                    Ok(adv) => {
+                        let sink = sink.clone();
+                        let name = adv
+                            .url
+                            .path_segments()
+                            .and_then(|s| s.last())
+                            .unwrap_or_else(|| adv.url.path());
+                        match client
                             .post(sink)
                             .header(header::CONTENT_TYPE, "application/json")
-                            .body(data.clone())
+                            .body(serde_json::to_string(&adv.csaf.clone()).unwrap())
                             .inject_token(&provider)
                             .await?
                             .send()
                             .await
                         {
                             Ok(r) if r.status() == StatusCode::CREATED => {
-                                log::info!(
-                                    "VEX ({}) of size {} stored successfully",
-                                    doc.document.tracking.id,
-                                    &data[..].len()
-                                );
+                                log::info!("VEX ({}) stored successfully", &adv.csaf.document.tracking.id);
                             }
                             Ok(r) => {
                                 log::warn!("(Skipped) {name}: Error storing VEX: {}", r.status());
                             }
                             Err(e) => {
```

Comment:
I think we should split this up into two paths here: server errors and client errors. In case of a server error, we should re-try and fail the application at some point. A client error we should record as well and proceed.